third_party/move/move-vm/runtime/src/data_cache.rs

```diff
@@ -128,108 +199,241 @@ impl TransactionDataCache {
+                resource_resolver,
                 addr,
-                &struct_tag,
-                &metadata,
-                layout_with_delayed_fields.layout_when_contains_delayed_fields(),
-            )?
+                struct_tag,
+                layout,
+                contains_delayed_fields,
+                load_data,
+            )?;
+            let num_bytes_loaded = if !existed_before {
+                NumBytes::new(bytes_loaded)
+            } else {
+                NumBytes::zero()
+            };
+            if let DataCacheGasMeterWrapper::Full(full_gas_meter) = gas_meter {
+                if let Some(v) = cached_info.maybe_value().and_then(|v| v.view()) {
+                    full_gas_meter.charge_deserialize_resource(
+                        *addr,
+                        &TypeWithRuntimeEnvironment {
+                            ty,
+                            runtime_environment: module_storage.runtime_environment(),
+                        },
+                        v,
+                    )?;
+                }
+                if !existed_before {
+                    full_gas_meter.charge_load_resource_bytes(
+                        *addr,
+                        &TypeWithRuntimeEnvironment {
+                            ty,
+                            runtime_environment: module_storage.runtime_environment(),
+                        },
+                        cached_info.exists()?,
+                        num_bytes_loaded,
+                    )?;
+                }
+            }
+            Ok((cached_info, num_bytes_loaded))
         };
 
-        let function_value_extension = FunctionValueExtensionAdapter { module_storage };
-        let (layout, contains_delayed_fields) = layout_with_delayed_fields.unpack();
-        let value = match data {
-            Some(blob) => {
-                let max_value_nest_depth = function_value_extension.max_value_nest_depth();
-                let val = ValueSerDeContext::new(max_value_nest_depth)
-                    .with_func_args_deserialization(&function_value_extension)
-                    .with_delayed_fields_serde()
-                    .deserialize(&blob, &layout)
-                    .ok_or_else(|| {
-                        let msg = format!(
-                            "Failed to deserialize resource {} at {}!",
-                            struct_tag.to_canonical_string(),
-                            addr
-                        );
-                        PartialVMError::new(StatusCode::FAILED_TO_DESERIALIZE_RESOURCE)
-                            .with_message(msg)
-                    })?;
-                GlobalValue::cached(val)?
+        let (entry, bytes_loaded) = match existing_entry {
+            Entry::Vacant(vacant_entry) => {
+                // Nothing is cached: charge for size and potentially bytes
+                let struct_tag = match module_storage.runtime_environment().ty_to_ty_tag(ty)? {
```

Comment:
this whole match could be a method either on environment that internally expects struct, or actually it could be a method on TypeWithRuntimeEnvironment (ty_to_struct_tag or struct_ty_to_tag or smt similar) - in which case we should also look at other uses of TypeWithRuntimeEnvironment maybe they are all for structs and could get rid of direct ty_to_ty_tag forwarding.

But see another comment about TypeWithRuntimeEnvironment - now it's defined twice.